src/Type.hs

```diff
@@ -135,22 +140,70 @@ splitBehaviour store (Constructor name contract iface@(Interface _ decls) iffs (
   ensures <- mapM (checkBool env) (fromMaybe [] maybeEnsures)
   let postcs = ensures <> mkStorageBounds store stateUpdates
 
+  let ethEnvs' = concat $ ethEnvFromExp <$> concat [iffs', invariants, ensures]
+      ethEnvs'' = concat $ ethEnvFromStorageUpdate <$> (rights stateUpdates)
+      ethEnvBounds = mkEthEnvBounds $ ethEnvs' <> ethEnvs''
+
   return $ ((I . (Invariant contract)) <$> invariants)
-           <> (splitCase name True contract iface (LitBool True) iffs' Nothing stateUpdates postcs)
+           <> (splitCase name True contract iface ethEnvBounds iffs' Nothing stateUpdates postcs)
 
 mkEnv :: Id -> Store -> [Decl]-> Env
 mkEnv contract store decls = (contract, fromMaybe mempty (Map.lookup contract store), store, abiVars)
  where
    abiVars = Map.fromList $ map (\(Decl typ var) -> (var, metaType typ)) decls
 
 -- | split case into pass and fail case
-splitCase :: Id -> Bool -> Id -> Interface -> Exp Bool -> [Exp Bool] -> Maybe ReturnExp
+splitCase :: Id -> Bool -> Id -> Interface -> [Exp Bool] -> [Exp Bool] -> Maybe ReturnExp
           -> [Either StorageLocation StorageUpdate] -> [Exp Bool] -> [Claim]
 splitCase name creates contract iface if' [] ret storage postcs =
-  [ B $ Behaviour name Pass creates contract iface if' (mconcat postcs) storage ret ]
+  [ B $ Behaviour name Pass creates contract iface (mconcat if') (mconcat postcs) storage ret ]
 splitCase name creates contract iface if' iffs ret storage postcs =
-  [ B $ Behaviour name Pass creates contract iface (mconcat (if':iffs)) (mconcat postcs) storage ret,
-    B $ Behaviour name Fail creates contract iface (And if' (Neg (mconcat iffs))) (mconcat postcs) storage Nothing ]
+  [ B $ Behaviour name Pass creates contract iface (mconcat (if' <> iffs)) (mconcat postcs) storage ret,
+    B $ Behaviour name Fail creates contract iface (And (mconcat if') (Neg (mconcat iffs))) (mconcat postcs) storage Nothing ]
+
+mkEthEnvBounds :: [EthEnv] -> [Exp Bool]
+mkEthEnvBounds vars = catMaybes $ mkBound <$> nub vars
+  where
+    mkBound :: EthEnv -> Maybe (Exp Bool)
```

Comment:
why not do this in terms of `bound` instead?

Reply:
much cleaner, even helped me find a little bug :pray: :sparkles: